README.md
# ares
Finding areas of convex polygon masks in images


ares/CHANGELOG.md
# Revision history for ares

## 0.1.0.0 -- YYYY-mm-dd

* First version. Released on an unsuspecting world.


arej/main.jl
# requires LinearAlgebra, Images, ImageView

include("GrahamScan.jl")

using ImageView, Images
using SparseArrays
using Random
using LinearAlgebra
using Statistics

function FindRoughEdge(A)
    # Find the rough coordinates of the non-horizontal edges of shapes in A
    B = Float64.(A);
    abs.(B[:, 2:end] - B[:, 1:end - 1]) |>
        rough_outline -> findall(px -> abs.(px - 1.0) < 0.8, rough_outline)
end

function FitLinear(x, y, deg)
    # Use linear regression to approximate f(x) = y with a polynomial of degree deg
    # returns coefficients of polynomial
    vs = hcat(map(p -> x.^p, (deg:-1:0))...);
    Q, R = qr(vs);
    R_pad = [R; zeros(size(Q, 1) - size(R, 1), size(R, 2))];
    params = R_pad \ (Q' * y);
    params
end

function FitLinearModel(x, y, deg)
    # Use linear regression to approximate f(x) = y with a polynomial of degree deg
    # returns a ready to use function
    params = FitLinear(x, y, deg);
    t -> (t.^(deg:-1:0))' * params
end

function CleanUpEdges(sensitivity, contour)
    # Try to remove unnecessary points from contour
    # contour is a matrix where the rows are x,y coordinate pairs
    reduced_contour = [];
    current_xs = Array{eltype(contour),1}();
    current_ys = Array{eltype(contour),1}();
    for point in eachrow(contour) # the first element is always equal to GrahamScan.EdgePoint(contour)
        x, y = point[1], point[2]
        if isempty(current_xs)
            push!(current_xs, x);
            push!(current_ys, y);
        else
            next_xs = [x; current_xs];
            next_ys = [y; current_ys];
            # approximate all points in current segment by a line
            approx = FitLinearModel(next_xs, next_ys, 1);
            # calculate deviations from approximation and actual values
            # and find the maximum deviation
            md = vcat(approx.(next_xs)...) - next_ys |>
                ds -> abs.(ds) |>
                maximum;
            if md <= sensitivity # if the maximum deviation doesn't exceed `sensitivity` pixels
                # try the next line
                current_xs = next_xs;
                current_ys = next_ys;
                continue
            else
                # otherwise save ends of the last line to the reduced contour
                push!(reduced_contour, [current_xs[1], current_ys[1]]);
                push!(reduced_contour, [current_xs[end], current_ys[end]]);
                # and begin a new segment starting at the current point
                current_xs = [x];
                current_ys = [y];
            end
        end
    end
    reduced_contour |> unique |>
        x -> hcat(x...)'
end

function Center(C)
    # find "center" of a point cloud
    mean.(eachcol(C))'
end

function Triangulate(center, points)
    start = points[1, :];
    current = start;
    Channel() do channel
        for point in eachrow(points[2:end, :])
            put!(channel, (current', center, point'));
            current = point;
        end
        put!(channel, (current', center, start'));
    end
end

function CalculateArea(C)
    # calculate area of star domain with center Center(C)
    triangleArea((a, b, c)) = abs(det([a 1; b 1; c 1])) / 2;
    map(triangleArea, Triangulate(Center(C), C)) |> sum
end

function CartesianToCoords(idxs)
    # convert CartesianIndex array to (no of pixels) x (2) Matrix
    hcat(map(idx -> [idx[1], idx[2]], idxs)...)'
end

function CoordsToImage(A)
    S = sparse(A[:,1], A[:,2], ones(size(A)[1]));
    I, J, V = findnz(S);
    I = I .- (minimum(I) - 1);
    J = J .- (minimum(J) - 1);
    sparse(I, J, V) |> Matrix
end

function Show(A)
    imshow(CoordsToImage(A))
    A
end

if isempty(ARGS)
    norm_path = "norm.bmp";
    img_path = "in.bmp";
else
    norm_path = ARGS[1];
    img_path = ARGS[2];
end

# calculate area of a square we know to have an area of 1mm²
# (this could of course be replaced with a arbitrary shape of known area)
area_square = load(norm_path) |>
    img -> Gray.(img) |>
    FindRoughEdge |>
    CartesianToCoords |>
    GrahamScan.FindConvexHull |> Show |>
    CalculateArea

pixelsToMillimeters(x) = x / sqrt(area_square)

area(path) = load(path) |>
    img -> Gray.(img) |>
    FindRoughEdge |>
    CartesianToCoords |>
    GrahamScan.FindConvexHull |> Show |>
    hull -> CleanUpEdges(10, hull) |> Show |>
    pixelsToMillimeters |>
    CalculateArea |>
    area ->
        println("The are is $(area) mm². ~$(Int64(round(sqrt(area_square)))) pixels are 1 mm.")

area(img_path)

println("Press any button to continue...")
readline()

ares/plot.py

import matplotlib.pyplot as plt


# with open("sorted_set.txt", "r") as f:
#    sorted_set = f.read()
with open("convex_hull.txt", "r") as f:
    hull = f.read()
# with open("start.txt", "r") as f:
#    start = f.read()


def parse(string):
    """Parse string of shape "[[1,2],[3,4]]" to list of float"""
    lst = string.replace("[", "").replace("]", "").split(",")
    xs, ys = zip(*[[float(lst[i]), float(lst[i+1])]
                   for i in range(0, len(lst), 2)])
    return list(xs), list(ys)


#ps_x, ps_y = parse(sorted_set)
h_x, h_y = parse(hull)
#s_x, s_y = parse(start)

h_x.append(h_x[0])
h_y.append(h_y[0])

# for i, xy in enumerate(zip(ps_x, ps_y)):
#    plt.annotate(f"{i}", xy)
#plt.scatter(ps_x, ps_y)
plt.scatter(h_x, h_y)
plt.plot(h_x, h_y)
#plt.scatter(s_x, s_y)
#plt.annotate("start", list(zip(s_x, s_y))[0])
plt.gca().invert_yaxis()
plt.show()


arej/GrahamScan.jl
module GrahamScan

export FindConvexHull, EdgePoint, SortAngle

using LinearAlgebra

function EdgePoint(A)
    # find lexicographic minimum row of A with the second column having
    # preceeding importance over the first one.
    min_y = minimum(A[:,2]);
    candidates = A[:,2] .== min_y;
    min_x = minimum(A[candidates, 1]);
    [min_x min_y]
end

function SortAngle(points)
    function Cos(x,y)
        c = acos(x/sqrt(x^2 + y^2));
        if x > 0 && y > 0
            c
        elseif x < 0 && y > 0
            c + π/2
        elseif x > 0 && y < 0
            -c
        elseif x < 0 && y < 0
            -c - π/2
        else
            c
        end
    end
    f(p1,p2) = Cos(p1...) < Cos(p2...);
    sortslices(points; dims=1, lt=f)
end

function ToDirection(p1, p2, p3)
    T = [1 p1;
         1 p2;
         1 p3];
    d = det(T);
    if d < 0
        :right
    elseif d == 0
        :inline
    else
        :left
    end
end

function Scan(p1, ps)
    (number_of_points, _) = size(ps);
    if number_of_points < 2
        [p1; ps]
    else
        p2 = ps[1,:]';
        p3 = ps[2,:]';
        ps = ps[3:end,:];
        direction = ToDirection(p1, p2, p3);
        if direction == :right
            Scan(p1, [p3; ps]) # reject move, again with same tail and next head
        elseif direction == :left
            [p1; Scan(p2, [p3; ps])] # Save point, onward with next point
        elseif direction == :inline
            Scan(p1, [p3; ps]) # Discard mid point, onward with next point
        end
    end
end

function FindConvexHull(P)
    # P is a (number of points) x (2) matrix
    (number_of_points, _) = size(P);
    if number_of_points == 0
        return []
    else
        start = EdgePoint(P);
        start_idx = (P[:, 1] .== start[1]) .& (P[:, 2] .== start[2]);
        without_start = reshape(P[.~[start_idx start_idx]], :, 2);
        # transform p into a coordinate system where o is
        # the origin (relative to the previous system)
        TranslateTo(o, p) = p - o;
        TranslateCloud(o, cloud) = hcat(map(p -> TranslateTo(o, p), eachrow(cloud))...)';
        sorted_points = TranslateCloud(start', without_start) |>
            SortAngle |>
            c -> TranslateCloud(-start', c)
        if number_of_points <= 3
            return [start; sorted_points]
        else
            return Scan(start, sorted_points)
        end
    end
end

#=
Test 1
A = [1 -1; 3 4; -1 2; -1 -1]

GrahamScan(A) # should be A

Test 2
B = [2.3486669219560348 2.7215254125789414;
    2.4139648037583257  -4.681676398649306;
    -0.6260514711523584 6.850334341587001;
    -7.64216704719866   -2.7605815091225576;
    -5.026012253905725  1.5777744917255312;
    -6.4853594504513445 5.318092139833023;
    5.082500986696756   -1.896255262210346;
    -6.497957092016469  -6.549163826826357;
    7.051754797682392   -5.031079040862534;
    3.283138295813238   -8.076020743414723;
    7.805413024480348   9.26931407569144;
    7.910312138709568   5.417744228015254;
    2.4605192481055127  -1.4160149654176593;
    -2.877588474503572  -3.4053645090358486;
    0.5611514305387537  -9.743920605133937;
    8.555954343029967   -4.394948528913232;
    1.550287503925139   9.078244403761712;
    -3.271086970349943  1.934092461779306;
    -2.2965794125239603 1.2422580281368312;
    3.182950465475866   7.148107783065644]

C = [[0.5611514305387537,-9.743920605133937],[3.283138295813238,-8.076020743414723],[8.555954343029967,-4.394948528913232],[7.805413024480348,9.26931407569144],[3.182950465475866,7.148107783065644],[1.550287503925139,9.078244403761712],[-0.6260514711523584,6.850334341587001],[-6.4853594504513445,5.318092139833023],[-5.026012253905725,1.5777744917255312],[-7.64216704719866,-2.7605815091225576],[-6.497957092016469,-6.549163826826357]]
c = GrahamScan(B)
hcat(C...)' - c # should be 0
=#
    
end

ares/GrahamScan.hs
module GrahamScan (findConvexHull) where

import           Data.List

data Direction = Straight_ | Left_ | Right_ deriving (Show, Eq)

data Point a = Point a a deriving (Eq)

{- A -> B -> C
Transform Point B into a coordinate system where A is O
-}
transform :: Num a => Point a -> Point a -> Point a
transform (Point x1 y1) (Point x2 y2) = Point (x2 - x1) (y2 - y1)

instance Show a => Show (Point a) where
    show (Point x y) = "[" ++ show x ++ "," ++ show y ++ "]"

neg :: Num a => Point a -> Point a
neg (Point x y) = Point (-x) (-y)

invert :: Direction -> Direction
invert Left_  = Right_
invert Right_ = Left_
invert _      = Straight_


toDirection :: (Num a, Ord a) => Point a -> Point a -> Point a -> Direction
toDirection (Point x1 y1) (Point x2 y2) (Point x3 y3)
    | det < 0 = Right_
    | det > 0 = Left_
    | det == 0 = Straight_
    where det = (x2 - x1) * (y3 - y1) - (x3 - x1) * (y2 - y1)


toPath :: (Fractional a, Ord a) => [Point a] -> [Direction]
toPath (p1 : p2 : p3 : ps) = toDirection p1 p2 p3 : toPath (p2:p3:ps)
toPath _                   = []


safeEdge :: (Fractional a, Ord a) => Point a -> Point a -> Point a
safeEdge p1@(Point x1 y1) p2@(Point x2 y2)
    | y1 < y2 = p1
    | y1 > y2 = p2
    | x1 < x2 = p1
    | x1 > x2 = p2


sortAngle :: (RealFloat a, Ord a) => [Point a] -> [Point a]
sortAngle = sortBy f
    where
        f (Point x1 y1) (Point x2 y2) = compare t1 t2
            where 
                quadCorrAtan x y = case (x, y) of
                    (x, y)
                        | x > 0 && y > 0 -> t
                        | x < 0 -> pi + t
                        | otherwise -> 2 * pi + t
                        where t = atan $ y/x
                t1 = quadCorrAtan x1 y1
                t2 = quadCorrAtan x2 y2

grahamScan :: (RealFloat a, Ord a) => [Point a] -> [Point a]
grahamScan [] = []
grahamScan ps
    | length ps <= 3 = start : sortedSet
    | otherwise = scan start sortedSet
    where
        sortedSet = map (transform (neg start)) $ sortAngle $ map (transform start) $ delete start ps
        start = foldl1 safeEdge ps
        scan p1 (p2:p3:ps) = case toDirection p1 p2 p3 of
            Right_    -> scan p1 (p3:ps) -- reject move, again with same tail and next head
            Left_     -> p1 : scan p2 (p3:ps) -- Save point, onward with next point
            Straight_ -> scan p1 (p3:ps) -- Discard mid point, onward with next point
        scan p1 ps = p1:ps

findConvexHull :: (RealFloat a, Ord a) => [(a,a)] -> [(a,a)]
findConvexHull coords = map (\(Point a b) -> (a,b)) $ grahamScan ps
    where
        ps = map (uncurry Point) coords
        start = foldl1 safeEdge ps

main = do
    let b = [(2.3486669219560348, 2.7215254125789414), (2.4139648037583257, -4.681676398649306), (-0.6260514711523584, 6.850334341587001), (-7.64216704719866, -2.7605815091225576), (-5.026012253905725, 1.5777744917255312), (-6.4853594504513445, 5.318092139833023), (5.082500986696756, -1.896255262210346), (-6.497957092016469, -6.549163826826357), (7.051754797682392, -5.031079040862534), (3.283138295813238, -8.076020743414723), (7.805413024480348, 9.26931407569144), (7.910312138709568, 5.417744228015254), (2.4605192481055127, -1.4160149654176593), (-2.877588474503572, -3.4053645090358486), (0.5611514305387537, -9.743920605133937), (8.555954343029967, -4.394948528913232), (1.550287503925139, 9.078244403761712), (-3.271086970349943, 1.934092461779306), (-2.2965794125239603, 1.2422580281368312), (3.182950465475866, 7.148107783065644)]
    -- let b = [(8.55399098977255, -5.930116971532156), (8.993078735081951, 5.350288864081211), (6.462575470143449, 5.2023279494868095), (-0.5012082521461263, -2.7635646447002493), (3.751363445443568, 7.5960181467283086), (6.962166027728905, -9.191931082358213), (-8.52794838447116, -5.015800893866606), (1.0483771663661052, 9.428596406283297), (7.005750668172169, -2.448446435835619), (-3.5860638017359125, 0.9388330578258497), (-5.077761984368355, 6.195061385608756), (1.0511918777938334, 3.418444016286317), (1.910602883556649, 6.10140057355062), (8.664115383882425, 1.3982069922142841), (2.606953039007575, -5.4206314772628295), (-2.022582347655069, -1.2453738587345171), (-7.629138378656162, -0.9150031949537851), (-2.543767432130222, 0.6424452705761237), (-3.9156442101437205, 6.495791693992917), (8.991930799248248, 2.0626146636265226)]
    let ps = map (uncurry Point) b
    let start = foldl1 safeEdge ps
    let sortedSet = map (transform (neg start)) $ sortAngle $ map (transform start) $ delete start ps
    writeFile "start.txt" $ show [start]
    writeFile "sorted_set.txt" $ show sortedSet
    writeFile "convex_hull.txt" $ show $ grahamScan ps

